### ISSUE TITLE
L2ARC doesn't work with 4k-native AF disks

### ISSUE THREAD

**Issue by uzzzerrrname** [zfsonlinux/zfs#3893]
Hello.

I run CentOS 7.1 under Windows 10 Hyper-V (Gen.2 VM).
I tried to create a pool with ZIL and L2ARC using Windows Storage Spaces virtual disks passed-through to the Linux VM. Expected was an operational pool, but instead I had plenty of "Bad block number requested" messages on `/dev/tty0` referring to the disk I allocated for L2ARC.
I re-created the pool with `-o ashift=12` with the same result.
Adding (`zfs add`) a cache vdev after pool creation also doesn't work

An ordinary VHDX-backed disk works fine.

What's the difference? Well, here it is (sda is VHDX-backed, sdb is passed-through Storage Space)

```
# lsblk -td -o +SIZE,DISC-ALN,DISC-GRAN,DISC-MAX,DISC-ZERO
NAME ALIGNMENT MIN-IO OPT-IO PHY-SEC LOG-SEC ROTA SCHED    RQ-SIZE  RA WSAME  SIZE DISC-ALN DISC-GRAN DISC-MAX DISC-ZERO
sda          0   4096      0    4096     512    1 deadline     128 128    8M   16G        0       32M       4G         0
sdb          0   4096      0    4096    4096    1 noop         128 128  256M  160G        0        0B       0B         0
```

Since ZFS apparently supports 4k-native disks for data and logs, I think the issue is L2ARC-related.

**Comment by kernelOfTruth** [comment]
what ZoL version, release ?

**Comment by uzzzerrrname** [comment]
0.6.5.2, DKMS, from project's yum repo

**Comment by behlendorf** [comment]
If you still have the error messages can you post them in the issue.  "Bad block number requested" usually implies accessing beyond the end of device.  Perhaps it got the wrong capacity somehow?

**Comment by andrijapanic** [comment]
Hi,

is there any update on this, or plan to fix it, or workaround :)

Problem is also present on Ubuntu 16.04 where OpenZFS is supposed to be integrated in OS.

We have serious SSDs for L2ARC, but this bug renders them not usable (https://www.hgst.com/products/solid-state-solutions/ultrastar-ssd1600mm) 

I appreciate any inputs.

THanks.
Andrija

**Comment by behlendorf** [comment]
@andrijapanic could you post in the each command and error your getting when trying to use these devices.  Are you able to create to pool but latter get `Bad block number errors`?  What version of ZFS on Linux are you using?

**Comment by andrijapanic** [comment]
Currently we reinstalled the host with Ubuntu 14.04, ZfsOnLinux 0.6.5.7. 
But also same behaviour with Ubuntu 16.04, with builtin ZFS, as I understand openZFS for Linux (my colleague is on this, so not sure what was version there with Ubuntu 16.04)

As for commands, there are no issues if you use 4K HDD/SSD for pool itself (verified), but the issue starts happening (bad block number error) ONLY when you configure 4K HDD/SSD as L2ARC (no problems with ZIL, no problem with POOL itself).

Hope this helps.

ANyway, after we used HGST tool called Niagara to do re-format SSDs to 512e sector size, then no more errors and we can normally use these SSDs with ashift=12 for POOL (4K HGST HDD, 4K ZIL SSD, 4K L2ARC SSD)

thanks
